python/ray/tests/test_autoscaling_policy.py

```diff
@@ -0,0 +1,579 @@
+                node = Node(resources, join_immediately)
+                self.ip_to_nodes[ip] = node
+                if not join_immediately:
+                    join_time = self.virtual_time + self.node_startup_delay_s
+                    self.event_queue.put(
+                        Event(join_time, SIMULATOR_EVENT_NODE_JOINED, node)
+                    )
+
+    def submit(self, work):
+        if isinstance(work, list):
+            self.work_queue.extend(work)
+        else:
+            self.work_queue.append(work)
+
+    def _get_node_to_run(self, bundle, nodes):
+        for ip, node in nodes.items():
+            if node.bundle_fits(bundle):
+                return ip, node
+        return None, None
+
+    def _schedule_placement_group(self, pg, nodes):
+        # This scheduling algorithm is bad, but it is approximately as bad as
+        # the real placement group scheduler.
+        to_allocate = []
+        if (
+            pg.strategy == PlacementStrategy.STRICT_PACK
+            or pg.strategy == PlacementStrategy.PACK
+        ):
+            combined = collections.defaultdict(float)
+            for bundle in pg.bundles:
+                for k, v in bundle.items():
+                    combined[k] += v
+            ip, node_to_run = self._get_node_to_run(combined, nodes)
+            if node_to_run is None:
+                return False
+            to_allocate.append((combined, ip))
+        elif (
+            pg.strategy == PlacementStrategy.STRICT_SPREAD
+            or pg.strategy == PlacementStrategy.SPREAD
+        ):
+            # TODO (Alex): More accurate handling of non-STRICT_PACK groups.
+            remaining_nodes = nodes.copy()
+            for bundle in pg.bundles:
+                ip, node_to_run = self._get_node_to_run(bundle, remaining_nodes)
+                if node_to_run is None:
+                    return False
+                del remaining_nodes[ip]
+                to_allocate.append((bundle, ip))
+
+        for bundle, ip in to_allocate:
+            node = self.ip_to_nodes[ip]
+            node.allocate(bundle)
+        pg.start_time = self.virtual_time
+        end_time = self.virtual_time + pg.duration
+        self.event_queue.put(
+            Event(end_time, SIMULATOR_EVENT_PG_DONE, (pg, to_allocate))
+        )
+        if pg.start_callback:
+            pg.start_callback()
+        return True
+
+    def _schedule_task(self, task, nodes):
+        ip, node = self._get_node_to_run(task.resources, nodes)
+        if node is None:
+            return False
+
+        node.allocate(task.resources)
+        task.node = node
+        task.start_time = self.virtual_time
+        end_time = self.virtual_time + task.duration
+        self.event_queue.put(Event(end_time, SIMULATOR_EVENT_TASK_DONE, task))
+        if task.start_callback:
+            task.start_callback()
+        return True
+
+    def schedule(self):
+        # TODO (Alex): Implement a more realistic scheduling algorithm.
```

Comment:
By realistic, do you mean basically use the binpacking/resource demand scheduler?

Reply:
I mostly mean a scheduling policy more similar to raylet